### PR TITLE
Func oop

### PR DESCRIPTION
**Pull request by oneEyedCondor**

### CODE REVIEW

File: func/js/coffeemachine.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    return recipes[_currentMode].name;
+  };
+
+  this.__checkIngridients = function (recipe) {
+    return (
+      Number(this.__checkWaterAmount(recipe.water))
+      * Number(this.__checkMilkAmount(recipe.milk))
+      * Number(this.__checkGrindAmount(recipe.grind))
+    );
+  };
```
Comment (didevgen) [inline func/js/coffeemachine.js]: move the calculation to the separated fundtion

File: func/js/phone.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        values.find(value => {
+            if(typeof value != type) {
+                throw new Error(`${value} is not a ${type}`);
+            }
+        });
+        return true;
+    }
+
+    function _isValidName(value) {
+        let exp = new RegExp('^\\w{2,15}$');
```
Comment (didevgen) [inline func/js/phone.js]: always use regexps as a separated constants in higher scope

File: func/js/coffeemachine.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    if (value > _coffeMachineStorages[type] || value < 0) {
+      console.log(`${type} storage is ${_coffeMachineStorages[type]}, not ${value}`);
+      return false;
+    }
+    return true;
+  };
+
+  this.addAmountValidation = function (type, value) {
+    const sum = value + _ingridients[type];
+    if (sum > _coffeMachineStorages[type] || value < 0) {
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/coffeemachine.js]: fixed

File: func/js/coffeemachine.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  };
+
+  this.setMilk = function (value) {
+    if (!this.numberValidation(value)) return;
+    if (!this.amountValidation('milk', value)) return;
+    _ingridients.milk = value;
+  };
+
+  this.addMilk = function (milk) {
+    if (!this.numberValidation(milk)) return;
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/coffeemachine.js]: fixed

File: func/js/coffeemachine.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  this.setGrinding = function (grind) {
+    if (!this.numberValidation(grind)) return;
+    if (!this.amountValidation('grind', grind)) return;
+    _ingridients.grind = grind;
+  };
+
+  this.getMode = () => _currentMode;
+
+  this._stringValidation = function(mode) {
+    return /[A-z]/.test(mode);
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/coffeemachine.js]: fixed

File: func/js/coffeemachine.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  this._numberValidation = function(mode) {
+    return /[0-9]/.test(mode);
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/coffeemachine.js]: fixed

File: func/js/phone.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
+function Phone(name = 'noname', phoneBook = []) {
+    BasicDevice.apply(this, arguments);
+    let _call = false;
+    let _phoneBook = phoneBook;
+
+    function _isValidType(type, ...values) {
+        values.find(value => {
+            if(typeof value != type) {
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/phone.js]: fixed

File: func/js/phone.js (hunk 2 above)
Comment (oneEyedCondor) [reply func/js/phone.js]: fixed

File: func/js/phone.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    function _isUndefined(value, errorMessage) {
+        if (value === undefined) {
+            throw new Error(errorMessage);
+        }
+    }
+
+    function _findElem(prop, matchVal) {
+        return _phoneBook.find(contact => {
+            if(contact[prop] === matchVal) {
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/phone.js]: fixed

File: func/js/ui.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+const onOff = document.getElementById('onOff');
+
+const refresh = document.getElementById('refresh');
+
+const readyOutput = document.getElementsByClassName('ready-coffee')[0];
+const drinks = document.getElementsByClassName('real-coffee__drink');
+
+// Functions
+function onSetWatter() {
+  c.setWater(Number(waterInput.value));
```
[the remark this answers is not in the document]
Comment (oneEyedCondor) [reply func/js/ui.js]: fixed

File: es6/js/phone.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
+class Phone extends BasicDevice {
+  constructor(name = 'noname', phoneBook = []) {
+    super(name);
+    this._call = false;
+    this._phoneBook = phoneBook;
+  }
+
+  static get NAME_REGEX() {
+    return new RegExp('^\\w{2,15}$');
+  }
```
Comment (didevgen) [inline es6/js/phone.js]: as a variant :)
I meant just a 
export const NAME_REGEX = new RegExp('^\\w{2,15}$');

File: es6/js/phone.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    return new RegExp('^\\+\\d{12}$');
+  }
+
+  _isValidType(type, ...values) {
+    values.find(value => {
+      if (typeof value !== type) {
+        throw new Error(`${value} is not a ${type}`);
+      }
+    });
+    return true;
```
Comment (didevgen) [inline es6/js/phone.js]: return values.some()

You are interested in true/false only

File: es6/js/phone.js [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  _isUndefined(value, errorMessage) {
+    if (value === undefined) {
+      throw new Error(errorMessage);
+    }
+  }
+
+  _findElem(prop, matchVal) {
+    return this._phoneBook.find(contact => contact[prop] === matchVal);
```
Comment (didevgen) [inline es6/js/phone.js]: might be undefined :)

File: es6/js/phone.js [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+      throw new Error('Phone is off');
+    }
+  }
+
+  deleteContact(id) {
+    if (this._state && this._isValidType('number', id)) {
+      for (let index = this._phoneBook.length; --index >= 0; ) {
+        if (this._phoneBook[index].id === id) {
+          this._phoneBook.splice(index, 1);
+        }
```
Comment (didevgen) [inline es6/js/phone.js]: this.phoneBook = this.phoneBook.filter....

File: func/js/coffeemachine.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+function Cofeemachine(name, water, milk, grind) {
+  // eslint-disable-next-line no-undef
+  BasicDevice.call(this, name);
+
+  const _coffeMachineStorages = {
+    water,
+    milk,
+    grind
+  };
+  const _modes = ['capuchino', 'americano', 'espresso'];
```
Comment (didevgen) [inline func/js/coffeemachine.js]: this array will be re-create each time you call CoffeeMachine

File: func/js/phone.js [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  };
+
+  this.editContact = function(id, newName, newPhoneNumber) {
+    if (
+      this._state &&
+      _isValidType('number', id) &&
+      _isValidType('string', newName, newPhoneNumber) &&
+      _isValidName(newName) &&
+      _isValidPhoneNumber(newPhoneNumber)
```
Comment (didevgen) [inline func/js/phone.js]: Btw this functions must be app-specific and should be public and in another file

File: func/index.func.html [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <meta http-equiv="X-UA-Compatible" content="ie=edge">
+  <link rel="stylesheet" href="style/coffeemachine.css">
+  <title>Document</title>
+</head>
+<body>
+
+  <div class="real-coffee-wrapper">
```
[the remark this answers is not in the document]
Comment (letrylesha) [reply func/index.func.html]: fixed

[the remark this answers is not in the document]
Comment (letrylesha) [reply func/index.func.html]: fixed

File: func/js/coffeemachine.js (hunk 14 above)
Comment (letrylesha) [reply func/js/coffeemachine.js]: fixed

**Comment by didevgen** [comment]
I might have been wrong. When following a such approach - it is better to branch from another *work* branch because I can't see any difference between another pull request

**Pull request closed without merging**